### PR TITLE
Add edge detection in GPIO module for Linux

### PR DESCRIPTION
**Pull request by hs0225**
- detect rising, falling and both edge
- tested on raspberryPi

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

File: src/platform/linux/iotjs_module_gpio-linux.c [hunk 1]
```diff
@@ -45,6 +47,96 @@
...
+
+static void gpio_edge_detection_cb(void* data) {
+  int fd;
+  iotjs_gpio_t* gpio = (iotjs_gpio_t*)data;
+
+  if ((fd = gpio_get_value_fd(gpio)) < 0)
+    return;
+
+  if (!gpio_clear_dummy_value(fd))
+    return;
```
Comment (daeyeon) [inline src/platform/linux/iotjs_module_gpio-linux.c]: no error log exists.

File: src/platform/linux/iotjs_module_gpio-linux.c [hunk 2]
```diff
@@ -45,6 +47,96 @@
...
+  return ret;
+}
+
+
+static void gpio_edge_detection_cb(void* data) {
+  int fd;
+  iotjs_gpio_t* gpio = (iotjs_gpio_t*)data;
+
+  if ((fd = gpio_get_value_fd(gpio)) < 0)
+    return;
```
Comment (daeyeon) [inline src/platform/linux/iotjs_module_gpio-linux.c]: no error log exists.

File: test/run_pass/test_gpio_event.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+];
+
+testGpioInfo.forEach(function(info) {
+  var switchGpio = gpio.open({
+    pin: info.pin,
+    edge: info.edge,
+    direction: gpio.DIRECTION.IN
+  }, function() {
+    switchGpio.on('change', function() {
+      console.log('pin:', info.pin, ', current value:', this.readSync());
```
Comment (daeyeon) [inline test/run_pass/test_gpio_event.js]: Assertion section is missing.

File: src/platform/linux/iotjs_module_gpio-linux.c [hunk 4]
```diff
@@ -45,6 +47,96 @@
...
+  }
+
+  return true;
+}
+
+
+static int gpio_edge_poll(int fd) {
+  int ret;
+  struct pollfd pollfd;
+
```
Comment (hs0225) [inline src/platform/linux/iotjs_module_gpio-linux.c]: I fixed it as your comments.

File: src/platform/linux/iotjs_module_gpio-linux.c (hunk 2 above)
Comment (hs0225) [inline src/platform/linux/iotjs_module_gpio-linux.c]: Done.

File: src/platform/linux/iotjs_module_gpio-linux.c [hunk 5]
```diff
@@ -45,6 +47,96 @@
...
+    return;
+
+  if (!gpio_clear_dummy_value(fd))
+    return;
+
+  while (true) {
+    if ((fd = gpio_get_value_fd(gpio)) < 0)
+      return;
+
+    if (gpio_edge_poll(fd) > 0)
```
Comment (hs0225) [inline src/platform/linux/iotjs_module_gpio-linux.c]: Done.

**Comment by zherczeg** [comment]
Shall we update the documentation?

**Comment by hs0225** [comment]
@zherczeg I'm going to update documentation after finishing implement on Nuttx.

File: src/modules/iotjs_module_gpio.h [hunk 6]
```diff
@@ -63,7 +71,12 @@ typedef struct {
   uint32_t pin;
   GpioDirection direction;
   GpioMode mode;
-#if defined(__TIZENRT__)
+  GpioEdge edge;
+#if defined(__linux__)
+  int value_fd;
+  uv_thread_t thread;
+  uv_mutex_t mutex;
```
Comment (LaszloLango) [inline src/modules/iotjs_module_gpio.h]: How big are `uv_mutex_t` and `uv_thread_t`? I think it would be better to store only pointers for them.

Comment (LaszloLango) [inline src/modules/iotjs_module_gpio.h]: Never mind, the struct is allocated with `IOTJS_ALLOC`, so it will be on the heap anyway.

**Pull request merged**